1. Project Title

## Customer Churn Analysis

2. Objective

## Project Objective
The objective of this project is to analyze customer churn data to identify the key factors influencing customer attrition. Using Python and Pandas,
the project performs data cleaning,exploratory data analysis(EDA) and business analysis to generate insights that can help improve customer retention and support data-driven business decisions.

3. Import Libraries

## Step 3: Import Required Libraries
Import the necessary Python libraries for data analysis.



In [1]:
import pandas as pd
import numpy as np

4. Load Dataset

## Step 4: Load Dataset
Read the Customer Churn Analysis CSV file into a Pandas Dataframe.

In [2]:
df = pd. read_csv(r"E:\downloads\WA_Fn-UseC_-Telco-Customer-Churn.csv")

5. First Five Records

## Step 5: Display the first five Records
Preview the first five rows to understand the dataset structure.

In [3]:
df.sample(5)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
1797,0839-QNXME,Female,0,No,No,30,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Credit card (automatic),19.40,578.5,No
5271,6507-DTJZV,Male,0,No,Yes,13,Yes,No,DSL,No,...,No,No,Yes,Yes,Month-to-month,No,Credit card (automatic),66.40,831.75,No
2895,7446-SFAOA,Female,0,Yes,No,37,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Bank transfer (automatic),19.85,717.5,No
2574,3269-ATYWD,Male,1,No,No,39,Yes,Yes,Fiber optic,No,...,No,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),95.65,3759.05,Yes
2712,5133-VRSAB,Male,0,No,No,8,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,No,Mailed check,29.35,216.45,No


6. Last Five Records

## Step 6: Display the Last Five Records
View the last five records of the dataset.



In [4]:
df.tail()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes
7042,3186-AJIEK,Male,0,No,No,66,Yes,No,Fiber optic,Yes,...,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No


7. Random Sample

## Step 7: Display Random 5 Records
Inspect randomly selected records for quick validation.

In [5]:
df.sample(5)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
3831,6946-LMSQS,Male,1,Yes,No,25,Yes,Yes,Fiber optic,Yes,...,No,No,No,Yes,One year,Yes,Electronic check,89.05,2177.45,Yes
285,6202-DYYFX,Female,0,No,No,22,Yes,Yes,Fiber optic,No,...,No,No,No,No,One year,Yes,Credit card (automatic),76.00,1783.6,No
2309,5414-OFQCB,Male,0,No,No,57,Yes,Yes,Fiber optic,No,...,No,Yes,Yes,No,Two year,Yes,Credit card (automatic),90.45,5229.8,No
374,6862-CQUMB,Male,0,No,No,37,Yes,No,Fiber optic,No,...,No,No,Yes,Yes,Month-to-month,Yes,Credit card (automatic),96.10,3646.8,No
6220,6434-TTGJP,Male,0,Yes,Yes,10,No,No phone service,DSL,Yes,...,Yes,Yes,No,Yes,Two year,No,Credit card (automatic),53.70,521,No


8. Dataset Shape

## Step 8: Check Dataset Shape
Find the number of rows and columns.



In [6]:
print(df.shape)

(7043, 21)


9. Column Names

## Step 9: Display Column Names
View all available columns in the dataset.

In [7]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')

10. Data Types

## Step 10: Check Data Types
Identify the datatype of every column.

In [8]:
df.dtypes

customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object

11. Complete Dataset Information

## Step 11: Dataset Information
Display summary information including non-null counts and memory usage.

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

12. Check Memory Usage

## Step 12: Check Memory Usage 
Check the memory usage of the dataset to understand its resource requirements.  

In [10]:
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

13. Explore Numerical Data

## Step 13: Statistical Summary of Numerical Columns
Generate a statistical summary of the numerical columns to understand their distribution and key characteristics.

In [135]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


14. Explore Categorical Data

## Step 14: Statistical Summary of Categorical Columns
Generate a statistical summary of the categorical columns to understand their unique values and frequency distribution.

In [136]:
df.describe(include="string")

,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,7590-VHVEG,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,20.2,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


15. Complete Dataset Summary 

## Step 15: Complete Statistical Summary  
Generate a complete statistical summary of the dataset to understand the distribution, central tendency, and key characteristics of both numerical and categorical columns.

In [137]:
df.describe(include="all")

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
count,7043,7043,7043.000000,7043,7043,7043.000000,7043,7043,7043,7043,...,7043,7043,7043,7043,7043,7043,7043,7043.000000,7043,7043
unique,7043,2,NaN,2,2,NaN,2,3,3,3,...,3,3,3,3,3,2,4,NaN,6531,2
top,7590-VHVEG,Male,NaN,No,No,NaN,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,NaN,20.2,No
freq,1,3555,NaN,3641,4933,NaN,6361,3390,3096,3498,...,3095,3473,2810,2785,3875,4171,2365,NaN,11,5174
mean,NaN,NaN,0.162147,NaN,NaN,32.371149,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,64.761692,NaN,NaN
std,NaN,NaN,0.368612,NaN,NaN,24.559481,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,30.090047,NaN,NaN
min,NaN,NaN,0.000000,NaN,NaN,0.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,18.250000,NaN,NaN
25%,NaN,NaN,0.000000,NaN,NaN,9.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35.500000,NaN,NaN
50%,NaN,NaN,0.000000,NaN,NaN,29.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,70.350000,NaN,NaN
75%,NaN,NaN,0.000000,NaN,NaN,55.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,89.850000,NaN,NaN


16. Check Data Quality

## Step 16: Check Missing Values
Check for missing values in each column to identify incomplete or unavailable data in the dataset.

In [138]:
df. isnull(). sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

17. Assess Missing Data

## Step 17: Calculate Missing Value Percentage
Calculate the percentage of missing values in each column to assess the extent of incomplete data in the dataset.

In [139]:
(df. isnull(). sum()/len(df))*100

customerID          0.0
gender              0.0
SeniorCitizen       0.0
Partner             0.0
Dependents          0.0
tenure              0.0
PhoneService        0.0
MultipleLines       0.0
InternetService     0.0
OnlineSecurity      0.0
OnlineBackup        0.0
DeviceProtection    0.0
TechSupport         0.0
StreamingTV         0.0
StreamingMovies     0.0
Contract            0.0
PaperlessBilling    0.0
PaymentMethod       0.0
MonthlyCharges      0.0
TotalCharges        0.0
Churn               0.0
dtype: float64

18. Assess Missing Data

## Step 18: Check Total Missing Values
Calculate the total number of missing values in the dataset to assess the overall completeness of the data.

In [140]:
df. isnull(). sum(). sum()

np.int64(0)

19. Check Data Quality

## Step 19: Check Duplicate Records
Check for duplicate records in the dataset to identify repeated customer entries that may affect the accuracy of the analysis.



In [141]:
df.duplicated().sum()

np.int64(0)

20. Clean Duplicate Records

## Step 20: Remove Duplicates
Remove duplicate records from the dataset to ensure each customer entry is unique and avoid duplication in the analysis.

In [142]:
df=df.drop_duplicates()

21. Verify Data Quality

## Step 21: Verify Duplicates Removed
Verify that all duplicate records have been removed from the dataset to ensure each customer entry is unique.

In [143]:
df.duplicated().sum()

np.int64(0)

22. Explore Data Values

## Step 22: Check Unique Values in each Column
Check the unique values in each column to understand the distinct categories and values present in the dataset.

In [144]:
df.nunique()

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

23. Explore Categorical Values

## Step 23: Display Unique Values of each Categorical Column
Display the unique values of each categorical column to identify the distinct categories present in the dataset.

In [145]:
for column in df.select_dtypes(include="string").columns:
    
      print("=" * 50)
      print(column)      
      print(df[column].unique()  
    )


customerID
<StringArray>
['7590-VHVEG', '5575-GNVDE', '3668-QPYBK', '7795-CFOCW', '9237-HQITU',
 '9305-CDSKC', '1452-KIOVK', '6713-OKOMC', '7892-POOKP', '6388-TABGU',
 ...
 '9767-FFLEM', '0639-TSIQW', '8456-QDAVC', '7750-EYXWZ', '2569-WGERO',
 '6840-RESVB', '2234-XADUH', '4801-JZAZL', '8361-LTMKD', '3186-AJIEK']
Length: 7043, dtype: str
gender
<StringArray>
['Female', 'Male']
Length: 2, dtype: str
Partner
<StringArray>
['Yes', 'No']
Length: 2, dtype: str
Dependents
<StringArray>
['No', 'Yes']
Length: 2, dtype: str
PhoneService
<StringArray>
['No', 'Yes']
Length: 2, dtype: str
MultipleLines
<StringArray>
['No phone service', 'No', 'Yes']
Length: 3, dtype: str
InternetService
<StringArray>
['DSL', 'Fiber optic', 'No']
Length: 3, dtype: str
OnlineSecurity
<StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str
OnlineBackup
<StringArray>
['Yes', 'No', 'No internet service']
Length: 3, dtype: str
DeviceProtection
<StringArray>
['No', 'Yes', 'No internet service']
Length: 3,

24. Analyze Categorical Distributions

## Step 24: Check Value Counts of each Categorical Column
Check the frequency of each unique category in the categorical columns to understand the distribution of customers across different categories.

In [146]:
for column in df.select_dtypes(include="string").columns:
    
      print("=" * 50)
      print(column)      
      print(df[column].value_counts())  
   

customerID
customerID
7590-VHVEG    1
5575-GNVDE    1
3668-QPYBK    1
7795-CFOCW    1
9237-HQITU    1
             ..
6840-RESVB    1
2234-XADUH    1
4801-JZAZL    1
8361-LTMKD    1
3186-AJIEK    1
Name: count, Length: 7043, dtype: int64
gender
gender
Male      3555
Female    3488
Name: count, dtype: int64
Partner
Partner
No     3641
Yes    3402
Name: count, dtype: int64
Dependents
Dependents
No     4933
Yes    2110
Name: count, dtype: int64
PhoneService
PhoneService
Yes    6361
No      682
Name: count, dtype: int64
MultipleLines
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64
InternetService
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64
OnlineSecurity
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64
OnlineBackup
OnlineBackup
No                     3088
Yes                    2429
No intern

25. Clean Text Data

## Step 25: Remove Leading and Trailing Spaces
Remove leading and trailing spaces from text values to ensure consistency and accuracy in the dataset.

In [147]:
for column in df.select_dtypes(include="string").columns:
    df[column]=df[column].str.strip()
     
     
   

In [148]:
print("Leading and trailing spaces removed successfully.")

Leading and trailing spaces removed successfully.


26. Verify Data Completeness

## Step 26: Recheck Missing Values
Recheck the dataset for missing values after data cleaning to confirm that no missing values remain.

In [149]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

27. Convert Data Types

## Step 27: Convert TotalCharges to Numeric
Convert the total_charges column to a numeric data type so it can be used accurately for calculations and analysis.

In [150]:
df["TotalCharges"]=pd.to_numeric(df["TotalCharges"],errors="coerce")

28. Verify Data Types

## Step 28: Check Data Types Again
Recheck the data types of all columns after data cleaning to confirm that each column has the correct data type.

In [151]:
df.dtypes

customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges        float64
Churn                   str
dtype: object

29. Verify Data Completeness

## Step 29: Check Missing Values After Conversion
Check for missing values created during the conversion of total_charges to numeric format to identify and handle invalid or blank entries.

In [152]:
df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

30. Handle Missing Values in Total Charges

## Step 30: Fill Missing Values in Total Charges
Fill the missing values in total_charges to ensure the column is complete and ready for further analysis. 

In [153]:
df["TotalCharges"]=df["TotalCharges"].fillna(df["TotalCharges"].median())

31. Verify Data Completeness

## Step 31: Verify Missing Values
Verify the dataset again to confirm that all missing values have been successfully handled after data cleaning.

In [154]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

32. Final Dataset Overview 

## Step 32: Final Dataset Information
Display the final dataset information to verify its structure, column data types, and number of non-null values after data cleaning.

In [155]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

33. Analyze Churn Distribution

## Step 33: Check Churn Distribution
Check the distribution of churned and retained customers to understand the overall customer churn pattern in the dataset.

In [156]:
df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

34. Analyze Churn Percentage

## Step 34: Churn Percentage
Calculate the percentage of customers who have churned to measure the overall customer churn rate.

In [157]:
df["Churn"].value_counts(normalize=True)*100

Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

35. Create Annual Charges 

## Step 35: Create Annual Charges Column
Create an annual_charges column by calculating the yearly charges from the customer's monthly charges.

In [158]:
df["Annual Charges"]= df["MonthlyCharges"]*12

36. Create Tenure Group

## Step 36: Create Tenure Group
Create a tenure_group column by grouping customers into meaningful tenure ranges to analyze churn patterns across different customer tenure levels.

In [159]:
df["TenureGroup"]=pd.cut(df["tenure"],bins=[0,12,24,36,48,60,72],
                         labels=[
                             "0-12 Months",
                             "13-24 Months",
                              "25-36 Months",
                              "37-48 Months",
                              "49-60 Months",
                              "61-72 Months"])

37. Verify New Columns

## Step 37: Check New Column 
Check the newly created column to verify that it had been added correctly and contains the expected values.

In [160]:
df[["tenure","TenureGroup"]].head()

,tenure,TenureGroup
0,1,0-12 Months
1,34,25-36 Months
2,2,0-12 Months
3,45,37-48 Months
4,2,0-12 Months


38. Create High-Value Customer Segment

## Step 38: Create High Value Customer Column
Create a high_value_customer column to identify customers with higher-value charges for targeted churn analysis.

In [161]:
df["HighValueCustomer"]=np.where(df["MonthlyCharges"] >=80,"Yes","No") 

39. Verify High-Value Customer Column

## Step 39: Check New Column
Check the newly created high_value_customer column to verify that it had been added correctly and contains the expected values.

In [162]:
df["HighValueCustomer"].value_counts()

HighValueCustomer
No     4366
Yes    2677
Name: count, dtype: int64

40. Sort Customers By Monthly Charges

## Step 40: Sort By Monthly Charges
Sort the dataset by monthly_charges to identify customers with the lowest and highest monthly charges.

In [163]:
df.sort_values(by="MonthlyCharges",ascending=False)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
4586,7569-NMZYQ,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,Yes,Bank transfer (automatic),118.75,8672.45,No,1425.0,61-72 Months,Yes
2115,8984-HPEMB,Female,0,No,No,71,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,Yes,Electronic check,118.65,8477.60,No,1423.8,61-72 Months,Yes
4804,5734-EJKXG,Female,0,No,No,61,Yes,Yes,Fiber optic,Yes,...,Yes,One year,Yes,Electronic check,118.60,7365.70,No,1423.2,61-72 Months,Yes
3894,5989-AXPUC,Female,0,Yes,No,68,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,No,Mailed check,118.60,7990.05,No,1423.2,61-72 Months,Yes
5127,8199-ZLLSA,Male,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,Yes,One year,Yes,Bank transfer (automatic),118.35,7804.15,Yes,1420.2,61-72 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1156,0621-CXBKL,Female,0,No,No,53,Yes,No,No,No internet service,...,No internet service,Two year,No,Mailed check,18.70,1005.70,No,224.4,49-60 Months,No
6906,9945-PSVIP,Female,0,Yes,Yes,25,Yes,No,No,No internet service,...,No internet service,Two year,Yes,Mailed check,18.70,383.65,No,224.4,25-36 Months,No
6652,0827-ITJPH,Male,0,No,No,36,Yes,No,No,No internet service,...,No internet service,Two year,Yes,Credit card (automatic),18.55,689.00,No,222.6,25-36 Months,No
1529,9764-REAFF,Female,0,Yes,No,59,Yes,No,No,No internet service,...,No internet service,Two year,No,Bank transfer (automatic),18.40,1057.85,No,220.8,49-60 Months,No


41. Identify Top 10 Customers by Monthly Charges

# Step 41: Top 10 customers by Monthly Charges
Identify the top 10 customers with the highest monthly_charges to analyze high-value customers based on monthly billing.

In [164]:
df.nlargest(10,"MonthlyCharges") 

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
4586,7569-NMZYQ,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,Yes,Bank transfer (automatic),118.75,8672.45,No,1425.0,61-72 Months,Yes
2115,8984-HPEMB,Female,0,No,No,71,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,Yes,Electronic check,118.65,8477.60,No,1423.8,61-72 Months,Yes
3894,5989-AXPUC,Female,0,Yes,No,68,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,No,Mailed check,118.60,7990.05,No,1423.2,61-72 Months,Yes
4804,5734-EJKXG,Female,0,No,No,61,Yes,Yes,Fiber optic,Yes,...,Yes,One year,Yes,Electronic check,118.60,7365.70,No,1423.2,61-72 Months,Yes
5127,8199-ZLLSA,Male,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,Yes,One year,Yes,Bank transfer (automatic),118.35,7804.15,Yes,1420.2,61-72 Months,Yes
6118,9924-JPRMC,Male,0,No,No,72,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,Yes,Electronic check,118.20,8547.15,No,1418.4,61-72 Months,Yes
4610,2889-FPWRM,Male,0,Yes,No,72,Yes,Yes,Fiber optic,Yes,...,Yes,One year,Yes,Bank transfer (automatic),117.80,8684.80,Yes,1413.6,61-72 Months,Yes
3205,3810-DVDQQ,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,Yes,Bank transfer (automatic),117.60,8308.90,No,1411.2,61-72 Months,Yes
6768,9739-JLPQJ,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,No,Credit card (automatic),117.50,8670.10,No,1410.0,61-72 Months,Yes
4875,2302-ANTDP,Female,1,Yes,No,48,Yes,Yes,Fiber optic,Yes,...,Yes,Month-to-month,No,Electronic check,117.45,5438.90,Yes,1409.4,37-48 Months,Yes


42. Identify Lowest 10 Customers by Monthly Charges

## Step 42: Lowest 10 Monthly Charges  
Identify the 10 customers with the lowest monthly_charges to analyze customers with the lowest billing levels.

In [165]:
df.nsmallest(10,"MonthlyCharges") 

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
3719,6823-SIDFQ,Male,0,No,No,28,Yes,No,No,No internet service,...,No internet service,One year,No,Credit card (automatic),18.25,534.70,No,219.0,25-36 Months,No
1529,9764-REAFF,Female,0,Yes,No,59,Yes,No,No,No internet service,...,No internet service,Two year,No,Bank transfer (automatic),18.40,1057.85,No,220.8,49-60 Months,No
6652,0827-ITJPH,Male,0,No,No,36,Yes,No,No,No internet service,...,No internet service,Two year,Yes,Credit card (automatic),18.55,689.00,No,222.6,25-36 Months,No
1156,0621-CXBKL,Female,0,No,No,53,Yes,No,No,No internet service,...,No internet service,Two year,No,Mailed check,18.70,1005.70,No,224.4,49-60 Months,No
6906,9945-PSVIP,Female,0,Yes,Yes,25,Yes,No,No,No internet service,...,No internet service,Two year,Yes,Mailed check,18.70,383.65,No,224.4,25-36 Months,No
4634,9426-SXNHE,Female,0,No,No,2,Yes,No,No,No internet service,...,No internet service,Month-to-month,No,Bank transfer (automatic),18.75,53.15,No,225.0,0-12 Months,No
1654,2967-MXRAV,Male,0,Yes,Yes,1,Yes,No,No,No internet service,...,No internet service,One year,No,Mailed check,18.80,18.80,No,225.6,0-12 Months,No
1846,7369-TRPFD,Male,0,No,No,19,Yes,No,No,No internet service,...,No internet service,One year,No,Mailed check,18.80,279.20,No,225.6,13-24 Months,No
2658,7473-ZBDSN,Female,0,Yes,Yes,14,Yes,No,No,No internet service,...,No internet service,Two year,No,Mailed check,18.80,255.55,No,225.6,13-24 Months,No
4356,3806-YAZOV,Female,0,No,No,3,Yes,No,No,No internet service,...,No internet service,Month-to-month,No,Mailed check,18.80,56.00,No,225.6,0-12 Months,No


43. Filter Churned Customers

## Step 43: Filter Churned Customers
Filter the dataset to identify customers who have churned for focused churn analysis.

In [166]:
df[df["Churn"]=="Yes"]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
13,0280-XJGEX,Male,0,No,No,49,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1244.4,49-60 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7021,1699-HPSBG,Male,0,No,No,12,Yes,No,DSL,No,...,No,One year,Yes,Electronic check,59.80,727.80,Yes,717.6,0-12 Months,No
7026,8775-CEBBJ,Female,0,No,No,9,Yes,No,DSL,No,...,No,Month-to-month,Yes,Bank transfer (automatic),44.20,403.35,Yes,530.4,0-12 Months,No
7032,6894-LFHLY,Male,1,No,No,1,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,75.75,75.75,Yes,909.0,0-12 Months,No
7034,0639-TSIQW,Female,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,No,Month-to-month,Yes,Credit card (automatic),102.95,6886.25,Yes,1235.4,61-72 Months,Yes


44. Filter Retained Customers

## Step 44: Filter Retained Customers
Filter the dataset to identify customers who have not churned for focused retention analysis.

In [167]:
df[df["Churn"]=="No"]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,358.2,0-12 Months,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,One year,No,Mailed check,56.95,1889.50,No,683.4,25-36 Months,No
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,507.6,37-48 Months,No
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.40,No,1069.2,13-24 Months,Yes
7,6713-OKOMC,Female,0,No,No,10,No,No phone service,DSL,Yes,...,No,Month-to-month,No,Mailed check,29.75,301.90,No,357.0,0-12 Months,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7037,2569-WGERO,Female,0,No,No,72,Yes,No,No,No internet service,...,No internet service,Two year,Yes,Bank transfer (automatic),21.15,1419.40,No,253.8,61-72 Months,No
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,One year,Yes,Mailed check,84.80,1990.50,No,1017.6,13-24 Months,Yes
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,One year,Yes,Credit card (automatic),103.20,7362.90,No,1238.4,61-72 Months,Yes
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,Month-to-month,Yes,Electronic check,29.60,346.45,No,355.2,0-12 Months,No


45. Filter Female Churned Customers

## Step 45: Female Customers Who Churned
Filter the dataset to identify female customers who have churned for focused demographic churn analysis.

In [168]:
df[
    (df["gender"]=="Female") &
     (df["Churn"]=="Yes")
    ]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
18,4190-MFLUW,Female,0,Yes,Yes,10,Yes,No,DSL,No,...,No,Month-to-month,No,Credit card (automatic),55.20,528.35,Yes,662.4,0-12 Months,No
29,8773-HHUOZ,Female,0,No,Yes,17,Yes,No,DSL,No,...,Yes,Month-to-month,Yes,Mailed check,64.70,1093.10,Yes,776.4,13-24 Months,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7008,1980-KXVPM,Female,1,No,No,3,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Credit card (automatic),75.05,256.25,Yes,900.6,0-12 Months,No
7010,0723-DRCLG,Female,1,Yes,No,1,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,74.45,74.45,Yes,893.4,0-12 Months,No
7011,5482-NUPNA,Female,0,No,No,4,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,60.40,272.15,Yes,724.8,0-12 Months,No
7026,8775-CEBBJ,Female,0,No,No,9,Yes,No,DSL,No,...,No,Month-to-month,Yes,Bank transfer (automatic),44.20,403.35,Yes,530.4,0-12 Months,No


46. Filter Male Churned Customers

## Step 46: Male Customers Who Churned
Filter the dataset to identify male customers who have churned for focused demographic churn analysis.

In [169]:
df[
    (df["gender"]=="Male") &
     (df["Churn"]=="Yes")
    ]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
13,0280-XJGEX,Male,0,No,No,49,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1244.4,49-60 Months,Yes
20,8779-QRDMV,Male,1,No,No,1,No,No phone service,DSL,No,...,Yes,Month-to-month,Yes,Electronic check,39.65,39.65,Yes,475.8,0-12 Months,No
22,1066-JKSGK,Male,0,No,No,1,Yes,No,No,No internet service,...,No internet service,Month-to-month,No,Mailed check,20.15,20.15,Yes,241.8,0-12 Months,No
26,6467-CHFZW,Male,0,Yes,Yes,47,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.35,4749.15,Yes,1192.2,37-48 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7009,7703-ZEKEF,Male,0,No,No,23,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,81.00,1917.10,Yes,972.0,13-24 Months,Yes
7018,1122-JWTJW,Male,0,Yes,Yes,1,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Mailed check,70.65,70.65,Yes,847.8,0-12 Months,No
7021,1699-HPSBG,Male,0,No,No,12,Yes,No,DSL,No,...,No,One year,Yes,Electronic check,59.80,727.80,Yes,717.6,0-12 Months,No
7032,6894-LFHLY,Male,1,No,No,1,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,75.75,75.75,Yes,909.0,0-12 Months,No


47. Filter Long-Term Customers

## Step 47: Long-Term Customers 
Filter the dataset to identify customers with long-term tenure for focused retention analysis.

In [170]:
df[df["tenure"]>=60]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
9,6388-TABGU,Male,0,No,Yes,62,Yes,No,DSL,Yes,...,No,One year,No,Bank transfer (automatic),56.15,3487.95,No,673.8,61-72 Months,No
15,3655-SNQYZ,Female,0,Yes,Yes,69,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,No,Credit card (automatic),113.25,7895.15,No,1359.0,61-72 Months,Yes
17,9959-WOFKT,Male,0,No,Yes,71,Yes,Yes,Fiber optic,Yes,...,Yes,Two year,No,Bank transfer (automatic),106.70,7382.25,No,1280.4,61-72 Months,Yes
28,5248-YGIJN,Male,0,Yes,No,72,Yes,Yes,DSL,Yes,...,Yes,Two year,Yes,Credit card (automatic),90.25,6369.45,No,1083.0,61-72 Months,Yes
30,3841-NFECX,Female,1,Yes,No,71,Yes,Yes,Fiber optic,Yes,...,No,Two year,Yes,Credit card (automatic),96.35,6766.95,No,1156.2,61-72 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7028,9281-CEDRU,Female,0,Yes,No,68,Yes,No,DSL,No,...,No,Two year,No,Bank transfer (automatic),64.10,4326.25,No,769.2,61-72 Months,No
7034,0639-TSIQW,Female,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,No,Month-to-month,Yes,Credit card (automatic),102.95,6886.25,Yes,1235.4,61-72 Months,Yes
7037,2569-WGERO,Female,0,No,No,72,Yes,No,No,No internet service,...,No internet service,Two year,Yes,Bank transfer (automatic),21.15,1419.40,No,253.8,61-72 Months,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,One year,Yes,Credit card (automatic),103.20,7362.90,No,1238.4,61-72 Months,Yes


48. Filter High Paying Customers

## Step 48: High Paying Customers
Filter the dataset to identify customers with high monthly charges for focused high-value customer analysis.

In [171]:
df[df["MonthlyCharges"]>=90]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
12,8091-TTVAX,Male,0,Yes,No,58,Yes,Yes,Fiber optic,No,...,Yes,One year,No,Credit card (automatic),100.35,5681.10,No,1204.2,49-60 Months,Yes
13,0280-XJGEX,Male,0,No,No,49,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1244.4,49-60 Months,Yes
14,5129-JLPIS,Male,0,No,No,25,Yes,No,Fiber optic,Yes,...,Yes,Month-to-month,Yes,Electronic check,105.50,2686.05,No,1266.0,25-36 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7023,1035-IPQPU,Female,1,Yes,No,63,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,103.50,6479.40,No,1242.0,61-72 Months,Yes
7025,2823-LKABH,Female,0,No,No,18,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),95.05,1679.40,No,1140.6,13-24 Months,Yes
7034,0639-TSIQW,Female,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,No,Month-to-month,Yes,Credit card (automatic),102.95,6886.25,Yes,1235.4,61-72 Months,Yes
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,One year,Yes,Credit card (automatic),103.20,7362.90,No,1238.4,61-72 Months,Yes


49. Group Customers by Gender

## Step 49: Group By Gender
Group customers by gender to analyze the distribution of male and female customers.

In [173]:
df.groupby("gender")["MonthlyCharges"].mean()

gender
Female    65.204243
Male      64.327482
Name: MonthlyCharges, dtype: float64

 50. Group By Contract Type

## Step 50: Group By Contract type
Group customers by contract type to analyze the distribution of customers across different contract categories.

In [174]:
df.groupby("Contract")["MonthlyCharges"].mean()

Contract
Month-to-month    66.398490
One year          65.048608
Two year          60.770413
Name: MonthlyCharges, dtype: float64

51. Group By Internet Service

## Step 51: Group By Internet Service
Group customers by internet service to analyze the distribution of customers across different internet service categories.

In [175]:
df.groupby("InternetService")["MonthlyCharges"].mean()

InternetService
DSL            58.102169
Fiber optic    91.500129
No             21.079194
Name: MonthlyCharges, dtype: float64

52. Group By Payment Method

## Step 52: Group By Payment Method
Group customers by payment method to analyze the distribution of customers across different payment methods.

In [176]:
df.groupby("PaymentMethod")["MonthlyCharges"].mean()

PaymentMethod
Bank transfer (automatic)    67.192649
Credit card (automatic)      66.512385
Electronic check             76.255814
Mailed check                 43.917060
Name: MonthlyCharges, dtype: float64

53. Apply Multiple Aggregations 

## Step 53: Multiple Aggregations
Apply multiple aggregation functions to summarize key customer metrics and gain a comprehensive view of the dataset.

In [177]:
df.groupby("Contract").agg({
    "MonthlyCharges":
    ["mean","max","min"],
    "TotalCharges":["mean","sum"],
    "tenure":"mean"
})

MonthlyCharges                TotalCharges               \
                         mean     max    min         mean          sum   
Contract                                                                 
Month-to-month      66.398490  117.45  18.75  1369.254581  5305861.500   
One year            65.048608  118.60  18.25  3033.571606  4468450.975   
Two year            60.770413  118.75  18.40  3715.179027  6297228.450   

                   tenure  
                     mean  
Contract                   
Month-to-month  18.036645  
One year        42.044807  
Two year        56.735103

54. Group By Churn

## Step 54: Group By Churn
Group customers by churn status to analyze the distribution of churned and retained customers.


In [178]:
df.groupby("Churn")[["MonthlyCharges",
  "TotalCharges", "tenure"]].mean()


,MonthlyCharges,TotalCharges,tenure
Churn,,,
No,61.265124,2552.882494,37.569965
Yes,74.441332,1531.796094,17.979133


55. Create Pivot Table

## Step 55: Pivot Table
Create a pivot table to summarize and compare customer data across different categories.

In [179]:
pd.pivot_table(
    df,
    values="MonthlyCharges",
    index="Contract",
    columns="Churn",
    aggfunc="mean" 
)

Churn,No,Yes
Contract,,
Month-to-month,61.462635,73.019396
One year,62.508148,85.050904
Two year,60.012477,86.777083


56. Analyze Contract VS Churn

## Step 56: Crosstab Contract VS Churn
Create a cross-tabulation of contract type and churn status to compare churn across different contract categories.

In [180]:
pd.crosstab(
    df["Contract"],
     df["Churn"]
)    
        

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1307,166
Two year,1647,48


57. Analyze Gender VS Churn

## Step 57: Crosstab Gender VS Churn
Create a cross-tabulation of gender and churn status to compare churn across male and female customers.

In [181]:
pd.crosstab(
    df["gender"],
     df["Churn"]
)    
        

Churn,No,Yes
gender,,
Female,2549,939
Male,2625,930


In [ ]:
58: Analyze Payment Method VS Churn

## Step 58: Crosstab Payment Method VS Churn
Create a cross-tabulation of payment method and churn status to compare churn across different payment methods.

In [182]:
pd.crosstab(
    df["PaymentMethod"],
     df["Churn"]
)    

Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),1286,258
Credit card (automatic),1290,232
Electronic check,1294,1071
Mailed check,1304,308


In [ ]:
59. Analyze Charges By Tenure Group

## Step 59: Average Charges By Tenure Group
Calculate the average monthly charges for each tenure group to compare customer charges across different tenure levels.

In [183]:
df.groupby("TenureGroup")["MonthlyCharges"].mean()


TenureGroup
0-12 Months     56.172023
13-24 Months    61.357275
25-36 Months    65.575481
37-48 Months    66.318241
49-60 Months    70.550781
61-72 Months    75.952701
Name: MonthlyCharges, dtype: float64

60. Analyze Numerical Relationships

## Step 60: Correlation Matrix 
Generate a correlation matrix to measure the strength and direction of relationships between numerical variables.

In [184]:
df.corr(numeric_only=True)

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges,Annual Charges
SeniorCitizen,1.000000,0.016567,0.220173,0.102652,0.220173
tenure,0.016567,1.000000,0.247900,0.825464,0.247900
MonthlyCharges,0.220173,0.247900,1.000000,0.650864,1.000000
TotalCharges,0.102652,0.825464,0.650864,1.000000,0.650864
Annual Charges,0.220173,0.247900,1.000000,0.650864,1.000000


In [ ]:
61. Analyze Customer Base 

## Step 61 : Total Number Of Customers
Calculate the total number of customers to understand the overall customer base size.

In [185]:
df.shape[0]

7043

In [ ]:
62. Analyze Churned Customers

## Step 62 : Total Number Of Churned Customers-Yes
Calculate the total number of customers who have churned from the dataset.

In [186]:
df[df["Churn"]=="Yes"].shape[0]

1869

In [ ]:
63. Analyze Retained Customers

## Step 63 : Total Number Of Churned Customers-No
Calculate the total number of customers who have not churned and remain with the company.

In [187]:
df[df["Churn"]=="No"].shape[0]

5174

In [ ]:
64. Analyze Overall Churn Rate

## Step 64 : Overall Churn Rate(%)
Calculate the overall churn rate to measure the percentage of customers who have churned from the total customer base.

In [188]:
(df["Churn"].value_counts(normalize=True)*100).round(2)

Churn
No     73.46
Yes    26.54
Name: proportion, dtype: float64

65. Analyze Gender Distribution

## Step 65 : Gender-Wise Customer Count
Calculate the overall churn rate to measure the percentage of customers who have churned from the total customer base.

In [189]:
df["gender"].value_counts()

gender
Male      3555
Female    3488
Name: count, dtype: int64

In [ ]:
66. Analyze Churn by Gender

## Step 66 : Gender-Wise Churn Count
Calculate the number of churned customers in each gender category to compare churn across male and female customers.

In [190]:
df.groupby("gender")["Churn"].value_counts()

gender  Churn
Female  No       2549
        Yes       939
Male    No       2625
        Yes       930
Name: count, dtype: int64

In [ ]:
67. Analyze Churn Percentage by Gender

## Step 67 : Gender-Wise Churn Percentage
Calculate the churn percentage for each gender category to compare the proportion of churned customers between male and female customers.

In [191]:
pd.crosstab(
 df["gender"],
 df["Churn"],   
normalize="index")*100    



Churn,No,Yes
gender,,
Female,73.079128,26.920872
Male,73.839662,26.160338


In [ ]:
68. Analyze Senior Citizen VS Churn

## Step 68 : Senior Citizen VS Churn
Compare churn across senior and non-senior customers to understand differences in customer retention.

In [192]:
pd.crosstab(
 df["SeniorCitizen"],
 df["Churn"]
)

Churn,No,Yes
SeniorCitizen,,
0,4508,1393
1,666,476


In [ ]:
69. Analyze Partner VS Churn

## Step 69 : Partner VS Churn
Compare churn across customers with and without a partner to understand differences in customer retention.

In [193]:
pd.crosstab(
 df["Partner"],
 df["Churn"]
)

Churn,No,Yes
Partner,,
No,2441,1200
Yes,2733,669


In [ ]:
70. Analyze Dependents VS Churn

## Step 70 : Dependents VS Churn
Compare churn across customers with and without dependents to understand differences in customer retention.

In [194]:
pd.crosstab(
 df["Dependents"],
 df["Churn"]
)

Churn,No,Yes
Dependents,,
No,3390,1543
Yes,1784,326


In [ ]:
71. Analyze Contract type VS Churn

## Step 71 : Contract type VS Churn
Compare churn across different contract types to understand differences in customer retention.

In [195]:
pd.crosstab(
 df["Contract"],
 df["Churn"]
)

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1307,166
Two year,1647,48


In [ ]:
72. Analyze Internet Service VS Churn

## Step 72 : Internet Service  VS Churn
Compare churn across different internet service categories to understand differences in customer retention.

In [196]:
pd.crosstab(
 df["InternetService"],
 df["Churn"]
)

Churn,No,Yes
InternetService,,
DSL,1962,459
Fiber optic,1799,1297
No,1413,113


In [ ]:
73. Analyze Payment Method VS Churn

### Step 73 : Payment Method  VS Churn
Compare churn across different payment methods to understand differences in customer retention.

In [197]:
pd.crosstab(
 df["PaymentMethod"],
 df["Churn"]
)

Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),1286,258
Credit card (automatic),1290,232
Electronic check,1294,1071
Mailed check,1304,308


In [ ]:
74. Analyze Paperless Billing VS Churn

## Step 74 : Paperless Billing VS Churn
Compare churn across customers with and without paperless billing to understand differences in customer retention.

In [198]:
pd.crosstab(
 df["PaperlessBilling"],
 df["Churn"]
)

Churn,No,Yes
PaperlessBilling,,
No,2403,469
Yes,2771,1400


In [ ]:
75. Analyze Phone Service VS Churn

## Step 75 : Phone Service VS Churn
Compare churn across customers with and without phone service to understand differences in customer retention.

In [199]:
pd.crosstab(
 df["PhoneService"],
 df["Churn"]
)

Churn,No,Yes
PhoneService,,
No,512,170
Yes,4662,1699


In [ ]:
76. Analyze Multiple Lines VS Churn

## Step 76 : Multiple Lines VS Churn
Compare churn across customers with different multiple-line service options to understand differences in customer retention.

In [200]:
pd.crosstab(
 df["MultipleLines"], 
 df["Churn"]
)

Churn,No,Yes
MultipleLines,,
No,2541,849
No phone service,512,170
Yes,2121,850


In [ ]:
77. Analyze Online Security VS Churn

## Step 77: Online Security VS Churn
Compare churn across customers with and without online security to understand differences in customer retention.

In [201]:
pd.crosstab(
 df["OnlineSecurity"], 
 df["Churn"]
)

Churn,No,Yes
OnlineSecurity,,
No,2037,1461
No internet service,1413,113
Yes,1724,295


In [ ]:
78. Analyze Online Backup VS Churn

## Step 78: Online Backup VS Churn
Compare churn across customers with and without online backup to understand differences in customer retention.

In [202]:
pd.crosstab(
 df["OnlineBackup"], 
 df["Churn"]
)

Churn,No,Yes
OnlineBackup,,
No,1855,1233
No internet service,1413,113
Yes,1906,523


In [ ]:
79. Analyze Device Protection VS Churn

## Step 79: Device Protection VS Churn
Compare churn across customers with and without device protection to understand differences in customer retention.

In [203]:
pd.crosstab(
 df["DeviceProtection"], 
 df["Churn"]
)

Churn,No,Yes
DeviceProtection,,
No,1884,1211
No internet service,1413,113
Yes,1877,545


In [ ]:
80. Analyze Tech Support VS Churn

## Step 80: Tech Support VS Churn
Compare churn across customers with and without tech support to understand differences in customer retention.

In [204]:
pd.crosstab(
 df["TechSupport"], 
 df["Churn"]
)

Churn,No,Yes
TechSupport,,
No,2027,1446
No internet service,1413,113
Yes,1734,310


In [ ]:
81. Analyze StreamingTV VS Churn

## Step 81: StreamingTV VS Churn
Compare churn across customers with StreamingTV to understand differences in customer retention.

In [205]:
pd.crosstab(
 df["StreamingTV"], 
 df["Churn"]
)

Churn,No,Yes
StreamingTV,,
No,1868,942
No internet service,1413,113
Yes,1893,814


In [ ]:
82. Analyze Streaming Movies VS Churn

## Step 82: Streaming Movies VS Churn
Compare churn across customers with different streaming movie service options to understand differences in customer retention.

In [206]:
pd.crosstab(
 df["StreamingMovies"], 
 df["Churn"]
)

Churn,No,Yes
StreamingMovies,,
No,1847,938
No internet service,1413,113
Yes,1914,818


In [ ]:
83. Analyze Average Monthly Charges By Churn

## Step 83: Average Monthly Charges By Churn
Calculate the average monthly charges for churned and retained customers to compare their billing levels.

In [207]:
df.groupby("Churn")["MonthlyCharges"].mean()

Churn
No     61.265124
Yes    74.441332
Name: MonthlyCharges, dtype: float64

In [ ]:
84. Analyze Average Total Charges By Churn

## Step 84: Average Total Charges By Churn
Calculate the average total charges for churned and retained customers to compare their overall billing levels.

In [208]:
df.groupby("Churn")["TotalCharges"].mean()

Churn
No     2552.882494
Yes    1531.796094
Name: TotalCharges, dtype: float64

In [ ]:
85. Analyze Average Tenure By Churn

## Step 85: Average Tenure By Churn
Calculate the average tenure of churned and retained customers to compare their customer relationship duration.

In [209]:
df.groupby("Churn")["tenure"].mean()

Churn
No     37.569965
Yes    17.979133
Name: tenure, dtype: float64

In [ ]:
86. Analyze Average Monthly Charges By Gender

## Step 86: Average Monthly Charges By Gender
Calculate the average monthly charges for each gender category to compare their average billing levels.

In [210]:
df.groupby("gender")["MonthlyCharges"].mean()

gender
Female    65.204243
Male      64.327482
Name: MonthlyCharges, dtype: float64

In [ ]:
87. Analyze Average Total Charges By Gender

## Step 87: Average Total Charges By Gender
Calculate the average total charges for each gender category to compare their overall billing levels.

In [211]:
df.groupby("gender")["TotalCharges"].mean()

gender
Female    2281.921323
Male      2281.912616
Name: TotalCharges, dtype: float64

In [ ]:
88. Analyze Average Tenure By Gender

## Step 88: Average Tenure By Gender
Calculate the average tenure for each gender category to compare their average customer relationship duration.

In [212]:
df.groupby("gender")["tenure"].mean()

gender
Female    32.244553
Male      32.495359
Name: tenure, dtype: float64

In [ ]:
89. Analyze Contract-Wise Customer Count

## Step 89: Contract-Wise Customer Count
Calculate the number of customers in each contract type to understand the distribution of customers across different contracts.

In [213]:
df["Contract"].value_counts()

Contract
Month-to-month    3875
Two year          1695
One year          1473
Name: count, dtype: int64

In [ ]:
90. Analyze Contract-Wise Average Monthly Charges

## Step 90: Contract-Wise Average Monthly Charges
Calculate the average monthly charges for each contract type to compare the average billing levels across different contracts.

In [214]:
df.groupby("Contract")["MonthlyCharges"].mean()

Contract
Month-to-month    66.398490
One year          65.048608
Two year          60.770413
Name: MonthlyCharges, dtype: float64

In [ ]:
91. Analyze Contract-Wise Total Revenue

## Step 91: Contract-Wise Total Revenue
Calculate the total revenue generated by each contract type to compare revenue contribution across different contracts.

In [215]:
df.groupby("Contract")["TotalCharges"].sum()

Contract
Month-to-month    5305861.500
One year          4468450.975
Two year          6297228.450
Name: TotalCharges, dtype: float64

In [ ]:
92. Analyze Internet Service Customer Count

## Step 92: Internet Service Customer Count
Calculate the number of customers in each internet service category to understand the distribution of customers across different services.

In [216]:
df["InternetService"].value_counts()

InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

In [ ]:
93. Analyze Internet Service-Wise Average Monthly Charges

## Step 93: Internet Service-Wise Average Monthly Charges
Calculate the average monthly charges for each internet service category to compare the average billing levels across different services.

In [217]:
df.groupby("InternetService")["MonthlyCharges"].mean()

InternetService
DSL            58.102169
Fiber optic    91.500129
No             21.079194
Name: MonthlyCharges, dtype: float64

In [ ]:
94. Analyze Internet Service-Wise Total Revenue

## Step 94: Internet Service-Wise Total Revenue
Calculate the total revenue generated by each internet service category to compare revenue contribution across different services.

In [218]:
df.groupby("InternetService")["TotalCharges"].sum()

InternetService
DSL            5128398.225
Fiber optic    9923622.950
No             1019519.750
Name: TotalCharges, dtype: float64

In [ ]:
95. Analyze Payment Method Customer Count

## Step 95: Payment Method Customer Count
Calculate the number of customers using each payment method to understand the distribution of customers across different payment methods.

In [219]:
df["PaymentMethod"].value_counts()

PaymentMethod
Electronic check             2365
Mailed check                 1612
Bank transfer (automatic)    1544
Credit card (automatic)      1522
Name: count, dtype: int64

In [ ]:
96. Analyze Payment Method-Wise Total Revenue

## Step 96: Payment Method-Wise Total Revenue
Calculate the total revenue generated by each payment method to compare revenue contribution across different payment methods.

In [220]:
df.groupby("PaymentMethod")["TotalCharges"].sum()

PaymentMethod
Bank transfer (automatic)    4751074.850
Credit card (automatic)      4672990.825
Electronic check             4944903.250
Mailed check                 1702572.000
Name: TotalCharges, dtype: float64

In [ ]:
97. Analyze Payment Method-Wise Average Monthly Charges

## Step 97: Payment Method-Wise Average Monthly Charges
Calculate the average monthly charges for each payment method to compare the average billing levels across different payment methods.

In [221]:
df.groupby("PaymentMethod")["MonthlyCharges"].mean()

PaymentMethod
Bank transfer (automatic)    67.192649
Credit card (automatic)      66.512385
Electronic check             76.255814
Mailed check                 43.917060
Name: MonthlyCharges, dtype: float64

In [ ]:
98. Analyze Tenure Group-Wise Customer Count

## Step 98: Tenure Group-Wise Customer Count
Calculate the number of customers in each tenure group to understand the distribution of customers across different tenure levels.

In [222]:
df["TenureGroup"].value_counts().sort_index()

TenureGroup
0-12 Months     2175
13-24 Months    1024
25-36 Months     832
37-48 Months     762
49-60 Months     832
61-72 Months    1407
Name: count, dtype: int64

In [ ]:
99. Analyze Tenure Group-Wise Churn Count

## Step 99: Tenure Group-Wise Churn Count
Calculate the number of churned customers in each tenure group to compare churn across different customer tenure levels.

In [223]:
pd.crosstab(df["TenureGroup"],df["Churn"])

Churn,No,Yes
TenureGroup,,
0-12 Months,1138,1037
13-24 Months,730,294
25-36 Months,652,180
37-48 Months,617,145
49-60 Months,712,120
61-72 Months,1314,93


In [ ]:
100. Analyze Tenure Group-Wise Churn Percentage

## Step 100: Tenure Group-Wise Churn Percentage
Calculate the churn percentage for each tenure group to compare the proportion of churned customers across different tenure levels.

In [101]:
pd.crosstab(df["TenureGroup"],df["Churn"],normalize="index")*100

Churn,No,Yes
TenureGroup,,
0-12 Months,52.321839,47.678161
13-24 Months,71.289062,28.710938
25-36 Months,78.365385,21.634615
37-48 Months,80.971129,19.028871
49-60 Months,85.576923,14.423077
61-72 Months,93.390192,6.609808


In [ ]:
101. Analyze High Value Customer Count

## Step 101: High Value Customer Count
Calculate the number of high-value customers to understand the size of the high-value customer segment.

In [102]:
df["HighValueCustomer"].value_counts()

HighValueCustomer
No     4366
Yes    2677
Name: count, dtype: int64

In [ ]:
102. Analyze High Value Customer VS Churn

## Step 102: High Value Customer VS Churn
Compare churn across high-value and non-high-value customers to understand differences in customer retention.

In [103]:
pd.crosstab(df["HighValueCustomer"],df["Churn"])

Churn,No,Yes
HighValueCustomer,,
No,3407,959
Yes,1767,910


In [ ]:
103. Identify Top Customers by Total Charges

## Step 103: Top 10 Customers by Total Charges
Identify the top 10 customers with the highest total charges to analyze the highest-value customers based on overall billing.

In [104]:
df.nlargest(
    10,
    "TotalCharges"
)[
    [
        "customerID",
        "MonthlyCharges",
          "TotalCharges",
           "tenure",
            "Churn" 
    ]
    ]

,customerID,MonthlyCharges,TotalCharges,tenure,Churn
4610,2889-FPWRM,117.80,8684.80,72,Yes
4586,7569-NMZYQ,118.75,8672.45,72,No
6768,9739-JLPQJ,117.50,8670.10,72,No
5347,9788-HNGUT,116.95,8594.40,72,No
2187,8879-XUAHX,116.25,8564.75,71,No
6118,9924-JPRMC,118.20,8547.15,72,No
5558,0675-NCDYU,116.40,8543.25,72,No
2368,6650-BWFRT,117.15,8529.50,72,No
2603,0164-APGRB,114.90,8496.70,72,No
2025,1488-PBLJN,116.85,8477.70,72,No


In [ ]:
104. Identify Top Customers by Monthly Charges

## Step 104: Top 10 Customers by Monthly Charges
Identify the top 10 customers with the highest monthly charges to analyze high-value customers based on monthly billing.

In [105]:
df.nlargest(
    10,
    "MonthlyCharges"
)[
    [
        "customerID",
        "MonthlyCharges",
          "TotalCharges",
           "Contract"
          
    ]
    ]

,customerID,MonthlyCharges,TotalCharges,Contract
4586,7569-NMZYQ,118.75,8672.45,Two year
2115,8984-HPEMB,118.65,8477.60,Two year
3894,5989-AXPUC,118.60,7990.05,Two year
4804,5734-EJKXG,118.60,7365.70,One year
5127,8199-ZLLSA,118.35,7804.15,One year
6118,9924-JPRMC,118.20,8547.15,Two year
4610,2889-FPWRM,117.80,8684.80,One year
3205,3810-DVDQQ,117.60,8308.90,Two year
6768,9739-JLPQJ,117.50,8670.10,Two year
4875,2302-ANTDP,117.45,5438.90,Month-to-month


In [ ]:
105. Identify Long-Term Customers

## Step 105: Top 10 Longest Tenure Customers
Identify the top 10 customers with the longest tenure to analyze the most loyal and long-term customers.

In [106]:
df.nlargest(
    10,
    "tenure"
)[
    [
        "customerID",
         "tenure",
          "Contract",
           "Churn"
          
    ]
    ]

,customerID,tenure,Contract,Churn
28,5248-YGIJN,72,Two year,No
35,6234-RAAPL,72,Two year,No
59,5954-BDFSG,72,Two year,No
62,0526-SXDJP,72,Two year,No
94,9848-JQJTX,72,Two year,No
106,6728-DKUCO,72,One year,No
109,2848-YXSMW,72,Two year,No
127,6734-PSBAW,72,Two year,No
140,3146-MSEGF,72,Two year,No
167,5997-OPVFA,72,Two year,No


In [ ]:
106. Identify Customers with Lowest Monthly Charges 

## Step 106: Lowest 10 Monthly Charge Customers
Identify the 10 customers with the lowest monthly charges to analyze customers with the lowest billing levels.

In [107]:
df.nsmallest(
    10,
    "MonthlyCharges"
)[
    [
        "customerID",
        "MonthlyCharges",
         "Contract"
           
    ]
    ]

,customerID,MonthlyCharges,Contract
3719,6823-SIDFQ,18.25,One year
1529,9764-REAFF,18.40,Two year
6652,0827-ITJPH,18.55,Two year
1156,0621-CXBKL,18.70,Two year
6906,9945-PSVIP,18.70,Two year
4634,9426-SXNHE,18.75,Month-to-month
1654,2967-MXRAV,18.80,One year
1846,7369-TRPFD,18.80,One year
2658,7473-ZBDSN,18.80,Two year
4356,3806-YAZOV,18.80,Month-to-month


In [ ]:
107. Identify Customers with Lowest Total Charges 

## Step 107: Lowest 10 Total Charge Customers
Identify the 10 customers with the lowest total charges to analyze customers with the lowest overall billing levels.



In [108]:
df.nsmallest(
    10,
    "TotalCharges"
)[
    [
        "customerID",
        "TotalCharges",
        "Contract"
         
          
            
    ]
    ]

,customerID,TotalCharges,Contract
1654,2967-MXRAV,18.80,One year
1151,8992-CEUEN,18.85,Month-to-month
6489,9318-NKNFC,18.85,Month-to-month
4939,9975-SKRNR,18.90,Month-to-month
583,1423-BMPBQ,19.00,Month-to-month
1733,1015-OWJKI,19.05,Month-to-month
3110,6569-KTMDU,19.10,Month-to-month
4348,6121-VZNQB,19.10,Month-to-month
5336,9441-QHEVC,19.10,Month-to-month
367,7302-ZHMHP,19.15,Month-to-month


108. Analyze Customers Using Multi-Level Group By

## Step 108: Multi-Level Group By 
Group customers by multiple categories to analyze customer distribution and key metrics across different segments.

In [109]:
df.groupby(["Contract","Churn"]
          )[
           ["MonthlyCharges","TotalCharges"]].mean()

MonthlyCharges  TotalCharges
Contract       Churn                              
Month-to-month No          61.462635   1521.927590
               Yes         73.019396   1164.460574
One year       No          62.508148   2902.417578
               Yes         85.050904   4066.212048
Two year       No          60.012477   3665.133576
               Yes         86.777083   5432.363542

In [ ]:
109. Create Pivot Table

## Step 109: Pivot Table
Create a pivot table to summarize and compare customer data across different categories and metrics.

In [110]:
pd.pivot_table(
    df,
    values="MonthlyCharges",
    index="Contract",
    columns="PaymentMethod",
    aggfunc="mean"
)

PaymentMethod,Bank transfer (automatic),Credit card (automatic),Electronic check,Mailed check
Contract,,,,
Month-to-month,69.085399,67.678821,74.989486,46.050056
One year,67.484143,67.972362,79.095245,44.306380
Two year,65.013918,64.422117,84.335714,38.587304


In [ ]:
110. Select Required Columns Using .loc

## Step 110: Select Specific Columns Using .loc
Select specific columns using .loc to access and analyze only the required data from the dataset.

In [111]:
df.loc[:, ["customerID","gender","Contract","MonthlyCharges","Churn"]]

,customerID,gender,Contract,MonthlyCharges,Churn
0,7590-VHVEG,Female,Month-to-month,29.85,No
1,5575-GNVDE,Male,One year,56.95,No
2,3668-QPYBK,Male,Month-to-month,53.85,Yes
3,7795-CFOCW,Male,One year,42.30,No
4,9237-HQITU,Female,Month-to-month,70.70,Yes
...,...,...,...,...,...
7038,6840-RESVB,Male,One year,84.80,No
7039,2234-XADUH,Female,One year,103.20,No
7040,4801-JZAZL,Female,Month-to-month,29.60,No
7041,8361-LTMKD,Male,Month-to-month,74.40,Yes


In [ ]:
111. Select First 10 Rows Using .iloc

## Step 111: Select First 10 Rows Using .iloc
Select the first 10 rows using .iloc to access a specific range of records from the dataset.

In [112]:
df.iloc[:10]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,358.2,0-12 Months,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,One year,No,Mailed check,56.95,1889.50,No,683.4,25-36 Months,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,507.6,37-48 Months,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.40,No,1069.2,13-24 Months,Yes
7,6713-OKOMC,Female,0,No,No,10,No,No phone service,DSL,Yes,...,No,Month-to-month,No,Mailed check,29.75,301.90,No,357.0,0-12 Months,No
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
9,6388-TABGU,Male,0,No,Yes,62,Yes,No,DSL,Yes,...,No,One year,No,Bank transfer (automatic),56.15,3487.95,No,673.8,61-72 Months,No


In [ ]:
112: Select Specific Rows and Columns Using .iloc

## Step 112: Select Rows and Columns Using .iloc
Select specific rows and columns using .iloc to access a particular subset of data from the dataset.

In [113]:
df.iloc[:10, :5]

,customerID,gender,SeniorCitizen,Partner,Dependents
0,7590-VHVEG,Female,0,Yes,No
1,5575-GNVDE,Male,0,No,No
2,3668-QPYBK,Male,0,No,No
3,7795-CFOCW,Male,0,No,No
4,9237-HQITU,Female,0,No,No
5,9305-CDSKC,Female,0,No,No
6,1452-KIOVK,Male,0,No,Yes
7,6713-OKOMC,Female,0,No,No
8,7892-POOKP,Female,0,Yes,No
9,6388-TABGU,Male,0,No,Yes


In [ ]:
113: Filter Data Using .query()

## Step 113: Filter Using .query()
Filter the dataset using .query() to retrieve rows that meet specific conditions.

In [114]:
df.query("MonthlyCharges >=80")

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.40,No,1069.2,13-24 Months,Yes
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
12,8091-TTVAX,Male,0,Yes,No,58,Yes,Yes,Fiber optic,No,...,Yes,One year,No,Credit card (automatic),100.35,5681.10,No,1204.2,49-60 Months,Yes
13,0280-XJGEX,Male,0,No,No,49,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1244.4,49-60 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7025,2823-LKABH,Female,0,No,No,18,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),95.05,1679.40,No,1140.6,13-24 Months,Yes
7034,0639-TSIQW,Female,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,No,Month-to-month,Yes,Credit card (automatic),102.95,6886.25,Yes,1235.4,61-72 Months,Yes
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,One year,Yes,Mailed check,84.80,1990.50,No,1017.6,13-24 Months,Yes
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,One year,Yes,Credit card (automatic),103.20,7362.90,No,1238.4,61-72 Months,Yes


In [ ]:
114. Filter Churned Customers Using .query()

## Step 114: Filter Churned Customers Using .query()
Filter the dataset using .query() to identify customers who have churned.


In [115]:
df.query("Churn == 'Yes'")

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
13,0280-XJGEX,Male,0,No,No,49,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1244.4,49-60 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7021,1699-HPSBG,Male,0,No,No,12,Yes,No,DSL,No,...,No,One year,Yes,Electronic check,59.80,727.80,Yes,717.6,0-12 Months,No
7026,8775-CEBBJ,Female,0,No,No,9,Yes,No,DSL,No,...,No,Month-to-month,Yes,Bank transfer (automatic),44.20,403.35,Yes,530.4,0-12 Months,No
7032,6894-LFHLY,Male,1,No,No,1,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,75.75,75.75,Yes,909.0,0-12 Months,No
7034,0639-TSIQW,Female,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,No,Month-to-month,Yes,Credit card (automatic),102.95,6886.25,Yes,1235.4,61-72 Months,Yes


In [ ]:
115. Filter Data Using Multiple Conditions

## Step 115: Filter Using Multiple Conditions
Filter the dataset using multiple conditions to identify customers who meet specific criteria.

In [116]:
df.query("Contract =='Month-to-month' and Churn == 'Yes'")

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
8,7892-POOKP,Female,0,Yes,No,28,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1257.6,25-36 Months,Yes
13,0280-XJGEX,Male,0,No,No,49,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,1244.4,49-60 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7018,1122-JWTJW,Male,0,Yes,Yes,1,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Mailed check,70.65,70.65,Yes,847.8,0-12 Months,No
7026,8775-CEBBJ,Female,0,No,No,9,Yes,No,DSL,No,...,No,Month-to-month,Yes,Bank transfer (automatic),44.20,403.35,Yes,530.4,0-12 Months,No
7032,6894-LFHLY,Male,1,No,No,1,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,75.75,75.75,Yes,909.0,0-12 Months,No
7034,0639-TSIQW,Female,0,No,No,67,Yes,Yes,Fiber optic,Yes,...,No,Month-to-month,Yes,Credit card (automatic),102.95,6886.25,Yes,1235.4,61-72 Months,Yes


In [ ]:
116. Filter Data Using .isin()

## Step 116: Filter Using .isin()
Filter the dataset using .isin() to select rows that match any value from a specified list of categories.

In [117]:
df[df["Contract"].isin(["Month-to-month", "One year"])]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,358.2,0-12 Months,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,One year,No,Mailed check,56.95,1889.50,No,683.4,25-36 Months,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,507.6,37-48 Months,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7036,7750-EYXWZ,Female,0,No,No,12,No,No phone service,DSL,No,...,Yes,One year,No,Electronic check,60.65,743.30,No,727.8,0-12 Months,No
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,One year,Yes,Mailed check,84.80,1990.50,No,1017.6,13-24 Months,Yes
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,One year,Yes,Credit card (automatic),103.20,7362.90,No,1238.4,61-72 Months,Yes
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,Month-to-month,Yes,Electronic check,29.60,346.45,No,355.2,0-12 Months,No


In [ ]:
117. Filter Data Using .between()

## Step 117: Filter Using .between()
Filter the dataset using .between() to select rows where a numerical value falls within a specified range.


In [118]:
df[df["MonthlyCharges"].between(50,100)]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,Annual Charges,TenureGroup,HighValueCustomer
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,One year,No,Mailed check,56.95,1889.50,No,683.4,25-36 Months,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,646.2,0-12 Months,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,848.4,0-12 Months,No
5,9305-CDSKC,Female,0,No,No,8,Yes,Yes,Fiber optic,No,...,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,1195.8,0-12 Months,Yes
6,1452-KIOVK,Male,0,No,Yes,22,Yes,Yes,Fiber optic,No,...,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.40,No,1069.2,13-24 Months,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7033,9767-FFLEM,Male,0,No,No,38,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Credit card (automatic),69.50,2625.25,No,834.0,37-48 Months,No
7035,8456-QDAVC,Male,0,No,No,19,Yes,No,Fiber optic,No,...,No,Month-to-month,Yes,Bank transfer (automatic),78.70,1495.10,No,944.4,13-24 Months,No
7036,7750-EYXWZ,Female,0,No,No,12,No,No phone service,DSL,No,...,Yes,One year,No,Electronic check,60.65,743.30,No,727.8,0-12 Months,No
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,One year,Yes,Mailed check,84.80,1990.50,No,1017.6,13-24 Months,Yes


In [ ]:
118. Create Customer Segment Using .apply()

## Step 118: Create Customer Segment Using .apply()
Create a customer segment using .apply() to classify customers into meaningful groups based on specific conditions.

In [119]:
def customer_segment(charge):
    if charge>=80:
        return "Premium"
    elif charge>=40:
         return "Standard"
    else:    
          return "Basic"
df["CustomerSegment"] =df["MonthlyCharges"].apply(customer_segment)        
        

In [ ]:
119. Verify Customer Segment

## Step 119: Check Customer Segment 
Check the customer segment column to verify that customers have been classified into the expected segments. 

In [120]:
df["CustomerSegment"].value_counts()

CustomerSegment
Premium     2677
Standard    2529
Basic       1837
Name: count, dtype: int64

In [ ]:
120. Rank Customers By Total Charges

## Step 120: Rank Customers By Total Charges
Rank customers based on their total charges to identify customers with the highest overall billing value.

In [121]:
df["Revenuerank"]=df["TotalCharges"].rank(ascending=False,method="dense")

In [ ]:
121. Convert Revenue Rank to Integer

## Step 121: Convert Revenue Rank to Integer
Convert the revenue_rank column to integer type for consistent ranking values.

In [122]:
df["revenue_rank"]=df["revenue_rank"]. astype(int)

KeyError: 'revenue_rank'

In [ ]:
122. Identify Top Customers by Revenue 

## Step 122: Top 20 Revenue Customers
Identify the top 20 customers with the highest revenue to analyze the customers contributing the most to overall revenue.

In [ ]:
df.nlargest(20, "TotalCharges")

In [ ]:
123. Add a New Column

## Step 123: Insert New Column
Insert a new company column at position 2 with the value telecom for all rows.

df.insert(
    2,"Company","Telecom")

In [ ]:
124. Rename Multiple Columns

## Step 124: Rename Multiple Columns
Rename multiple columns to improve column naming consistency and make the dataset easier to understand.

In [ ]:
df.rename(
    columns={"gender":"Gender","tenure":"Tenure"},inplace=True)

In [ ]:
125. Verify Column Names

## Step 125: Check Column Names 
Check the column names to verify that all columns have been named correctly after the changes. 

In [ ]:
df.columns

In [ ]:
126. Sort Data Using Multiple Columns

## Step 126: Sort Using Multiple Columns
Sort the dataset using multiple columns to organize and compare records based on more than one criterion.

In [ ]:
df.sort_values(
    by=["Contract","MonthlyCharges"],ascending=[True,False]
)

In [ ]:
127. Create Discount Column

## Step 127: Create Discount Column
Create a discount column to calculate and store discount values for each customer for further analysis.


In [ ]:
df["Discount"] =df["MonthlyCharges"] * 0.10


In [ ]:
128. Create Final Bil Column

## Step 128: Create final bill
Create a final_bill column to calculate the final bill amount for each customer after applying relevant discount.


In [ ]:
df["FinalBill"]=df["MonthlyCharges"]-df["Discount"]

In [ ]:
129. Map Churn Values

## Step 129: Map Churn Values
Map the churn values to meaningful numeric or categorical values to support analysis and modeling..

In [ ]:
df["ChurnFlag"]=df["Churn"].map(
    {
        "Yes": 1,
        "No":0
    }
)   

In [ ]:
130. Display Final Dataset

## Step 130: Display Final Dataset
Display the final dataset to verify its structure, values, and newly created columns after all data preparation steps.

In [ ]:
df.head()

In [ ]:
131. Replace Values

## Step 131: Replace Values
Replace existing values in the dataset with meaningful or standardized values to improve data consistency.

In [ ]:
df["Gender"]=df["Gender"].replace({"Male": "M", "Female": "F"})

In [ ]:
132. Create a New Column Using .assign()

## Step 132: Assign New Column
Assign a new column to the dataset to store additional information required for further analysis.

In [ ]:
df=df.assign(Revenue=df["MonthlyCharges"] * df["Tenure"])
    

In [ ]:
133. Filter Data Using .where()

## Step 133: Filter Using .where()
Filter the dataset using .where() to retain values that meet a specified condition.

In [ ]:
df["HighCharges"]=df["MonthlyCharges"].where(df["MonthlyCharges"] >=80, np.nan)

In [ ]:
134. Filter Data Using .mask()

## Step 134: Filter Using Mask
Filter the dataset using .mask() to replace or hide values that meet a specified condition.

In [ ]:
df["LowCharges"]=df["MonthlyCharges"].mask(df["MonthlyCharges"] >=80, np.nan)

In [ ]:
135. Create Charge Category

## Step 135: Create Charge Category
Create a charge_category column to classify customers into meaningful groups based on their monthly charges.

In [ ]:
df["ChargeCategory"] = pd.cut(df["MonthlyCharges"],bins=[0,35,70,120],labels=["Low","Medium","High"])
                              

In [ ]:
136. Analyze Customer Count By Charge Category

## Step 136: Customer Count By Charge Category
Calculate the number of customers in each charge category to understand the distribution of customers across different charge levels.

In [ ]:
df["ChargeCategory"].value_counts()

In [ ]:
137. Create Quartile-Based Segmentation

## Step 137: Quartile Segmentation
Divide customers into quartile-based segments to analyze customers across different revenue levels.

In [ ]:
df["RevenueQuartile"] = pd.qcut(df["TotalCharges"],q=4,labels=["Q1", "Q2" ,"Q3" ,"Q4"])

In [ ]:
138. Create Revenue Quartile Segmentation

## Step 138: Revenue Quartile Count
Calculate the number of customers in each revenue quartile to understand the distribution of customers across revenue segments.

In [ ]:
df["RevenueQuartile"].value_counts()

In [ ]:
139. Apply Multiple Aggregations

## Step 139: Multiple Aggregations
Apply multiple aggregation functions to summarize key customer metrics across different groups.

In [ ]:
df.groupby("Contract").agg(
Customer_Count=("customerID","count"),
Avg_Monthly_Charges=("MonthlyCharges","mean"),
Avg_Total_Charges=("TotalCharges","mean"), 
Max_Total_Charges=("TotalCharges","max"))    

140. Group Customers by Gender and Contract

## Step 140: Group By Gender and Contract
Group customers by gender and contract type to analyze customer distribution across these categories.

In [ ]:
df.groupby(["Gender","Contract"]).agg(
Customers=("customerID","count"),
AvgCharges=("MonthlyCharges","mean"))

141. Create Pivot Table

## Step 141: Pivot Table
Create a pivot table to summarize and compare customer data across gender and contract type.

In [ ]:
pd.pivot_table(
    df,
    values="MonthlyCharges",
    index="Gender",
    columns="Contract",
    aggfunc=["mean","max","min"])

142. Analyze Churn Rate By Contract

## Step 142: Churn Rate By Contract
Calculate the churn rate for each contract type to compare customer retention across different contracts.

In [ ]:
pd.crosstab(
    df["Contract"],df["Churn"],normalize="index").round(4) * 100

143. Analyze Churn Rate By Payment Method

## Step 143: Churn Rate By Payment Method
Calculate the churn rate for each payment method to compare customer retention across different payment options.

In [ ]:
pd.crosstab(
    df["PaymentMethod"],df["Churn"],normalize="index").round(4) * 100

144. Analyze Churn Rate By Internet Service

## Step 144: Churn Rate By Internet Service
Calculate the churn rate for each internet service category to compare customer retention across different services.

In [ ]:
pd.crosstab(
    df["InternetService"],df["Churn"],normalize="index").round(4) * 100

145. Change Pandas Column Names

## Step 145: Change The Pandas Column Names
Change The Pandas column names to improve naming consistency and make the dataset easier to work with.

In [ ]:
df.rename(columns={
    "customer ID": "customer_id",
    "Gender": "gender",
    "SeniorCitizen": "senior_citizen",
    "Partner": "partner",
    "Dependents": "dependents",
     "Tenure": "tenure",
     "PhoneService" : "phone_service",
      "MultipleLines": "multiple_lines",
       "InternetService": "internet_service",
      "OnlineSecurity": "online_security",
       "OnlineBackup": "online_backup",
       "DeviceProtection": "device_protection",
       "TechSupport": "tech_support",
       "StreamingTV": "streaming_tv",
       "StreamingMovies": "streaming_movies",
           "Contract": "contract",
           "PaperlessBilling": "paperless_billing",
       "PaymentMethod": "payment_method", 
           "MonthlyCharges": "monthly_charges",
           "TotalCharges": "total_charges",
           "Churn": "churn",
       "Annual Charges": "annual_charges",
           "TenureGroup": "tenure_group", 
           "HighValueCustomer": "high_value_customer",
           "CustomerSegment": "customer_segment",
       "Revenuerank": "revenue_rank",
           "Discount": "discount",
          "FinalBill": "final_bill", 
           "ChurnFlag": "churn_flag", 
           "Revenue": "revenue",
       "HighCharges": "high_charges",
           "LowCharges": "low_charges",
           "ChargeCategory": "charge_category",
           "RevenueQuartile": "revenue_quartile" }, inplace=True)
    
    

146. Fill Missing Values in Charge Columns

## Step 146: Fill Missing Values in Charge Columns
Fill missing values in high_charges and low_charges with 0 to ensure the charge-category columns are complete and ready for analysis.

In [ ]:
df["high_charges"]=df["high_charges"].fillna(0)
df["low_charges"]=df["low_charges"].fillna(0)

147. Visualize Churn Distribution

## Step 147: Churn Distribution(Count Plot)
Create a count plot to visualize the distribution of churned and retained customers.

In [124]:
import matplotlib.pyplot as plt
import seaborn as sns

sns.countplot(x="churn", data=df)
plt.title("Customer Churn Distribution")
plt.xlabel("Churn")
plt.ylabel("Number of Customers")
plt.show()

ValueError: Could not interpret value `churn` for `x`. An entry with this name does not appear in `data`.

## Business Insights
The majority of customers are retained, while a substantial majority have churned, highlighting a significant customer-retention challenge
and an opportunity to reduce customer loss through  targeted retention strategies.
 

148. Visualize Gender Distribution

## Step 148: Gender Distribution  
Create a count plot to visualize the distribution of customers across male and female categories.

In [ ]:
sns.countplot(x="gender", data=df)
plt.title("Gender Distribution")
plt.show()

## Business Insights
The customer base is fairly balanced between male and female customers, indicating that gender alone is unlikely to be a major factor in overall customer churn.


149. Visualize Churn by Gender

## Step 149: Churn by Gender 
Create a count plot to compare churned and retained customers across gender categories.

In [ ]:
sns.countplot(x="gender",hue="churn", data=df)
plt.title("Churn by Gender")
plt.xlabel("Gender")
plt.ylabel("Number of Customers")
plt.show()

## Business Insights
Male customers have a slightly higher churn count than female customers, but the difference is relatively small, indicating that gender alone does not appear to be a major driver of churn.    

150. Visualize Contract Type by Churn

## Step 150: Contract Type by Churn 
Create a count plot to compare churned and retained customers across different contract types.

In [ ]:
sns.countplot(x="contract",hue="churn", data=df)
plt.title("Churn by Contract Type") 
plt.xlabel("Contract Type")
plt.ylabel("Number of Customers")
plt.show()

## Business Insights
Customers on month-to-month contracts show substantially higher churn, while customers with longer-term contracts are more likely to remain,
highlighting contract duration as an important retention factor.

151. Visualize Internet Service by Churn

##  Step 151: Internet Service by Churn
Create a count plot to compare churned and retained customers across different internet service categories.

In [ ]:
sns.countplot(x="internet_service",hue="churn", data=df) 
plt.title("Churn by Internet Service") 
plt.show()

## Business Insights
Fiber optic customers show the highest churn, indicating that this segment may require targeted retention strategies to reduce customer loss.

152. Visualize Payment Method by Churn

## Step 152: Payment Method by Churn
Create a count plot to compare churned and retained customers across different payment methods.

In [ ]:
sns.countplot(x="payment_method",hue="churn", data=df) 
plt.title("Churn by Payment Method") 
plt.xticks(rotation=30)
plt.show()

## Business Insights
Customers using electronic check have the highest number of churned customers, indicating  that this payment method may be associated
with higher customer churn and should be a focus for retention strategies.

153. Visualize Monthly Charges Distribution

## Step 153: Monthly Charges Distribution
Create a histogram to visualize the distribution of customers across different monthly charge levels.

In [ ]:
sns.histplot(df["monthly_charges"],bins=30)  
plt.title("Monthly Charges Distribution") 
plt.show()

## Business Insights
Monthly charges are widely distributed, with a large concentration of customers at the lower charge range and a broader distribution
across mid-to-high charge levels.

154. Visualize Monthly Charges vs Churn

## Step 154: Monthly Charges vs Churn
Create a box plot to compare monthly charges between churned and retained customers.

In [ ]:
sns.boxplot(x="churn",y="monthly_charges", data=df) 
plt.title("Monthly Charges by Churn") 
plt.show()

## Business Insights
Churned customers have higher monthly charges than non-churned customers, suggesting that higher monthly charges may be associated
with increased churn risk.

155. Visualize Tenure Distribution

## Step 155: Tenure Distribution
Create a histogram to visualize the distribution of customers across different tenure levels.

In [ ]:
sns.histplot(df["tenure"],bins=30)  
plt.title("Tenure Distribution") 
plt.show()

## Business Insights
A substantial number of customers have relatively low tenure, highlighting the importance of strengthening early-stage retention 
strategies to reduce customer churn.

156. Visualize Tenure vs Churn

## Step 156: Tenure vs Churn
Create a box plot to compare customer tenure between churned and retained customers.

In [ ]:
sns.boxplot(x="churn",y="tenure", data=df) 
plt.title("Tenure by Churn") 
plt.show()

## Business Insights
Churned customers generally have shorter tenure than retained customers, highlighting the importance of strengthening early-stage
customer retention strategies.

157. Visualize Total Charges vs Churn

## Step 157: Total Charges vs Churn
Create a box plot to compare total charges between churned and retained customers.


In [ ]:
sns.boxplot(x="churn",y="total_charges", data=df) 
plt.title("Total Charges by Churn") 
plt.show()

## Business Insights
Churned customers generally have lower total charges, which is consistent with their shorter tenure and indicates an opportunity to increase customer lifetime value through improved retention.

158. Visualize Senior Citizen vs Churn

## Step 158: Senior Citizen vs Churn
Create a count plot to compare churned and retained customers across senior citizen status.


In [ ]:
sns.countplot(x="senior_citizen",hue="churn", data=df) 
plt.title("Senior Citizen vs Churn") 
plt.show()

## Business Insights
Senior customers show a somewhat higher churn proportion than non-senior customers, suggesting that this segment may benefit from
personalized support and retention offers.

159. Visualize Correlations

## Step 159: Correlation Heatmap
Create a heatmap to visualize the strength and direction of relationships between numerical variables.

In [ ]:
plt.figure(figsize=(8,6))
sns.heatmap(df.corr(numeric_only=True), annot=True, cmap="Blues")
plt.title("Correlation Heatmap") 
plt.show()

## Business Insights
Tenure has the strongest relationship with churn, showing a negative correlation of approximately -0.35, suggesting that customers
with shorter tenure are more likely to churn.

160. Visualize Numerical Relationships

## Step 160: Pair Plot
Create a pair plot to visualize relationships and distributions among selected numerical variables by churn status.

In [ ]:
sns.pairplot(
df[["tenure","monthly_charges","total_charges","churn"]], hue="churn")
plt.show()
                  

## Business Insights
Churned customers are more concentrated among customers with shorter tenure and lower total charges, while longer-tenure
customers are more likely to remain.

161. Visualize Churn Distribution

## Step 161: Pie Chart of Churn
Create a pie chart to visualize the proportion of churned and retained customers.

In [123]:
df["churn"].value_counts().plot(kind="pie",autopct="%1.1f%%")
plt.title("Customer Churn Percentage") 
plt.ylabel(" ")
plt.show()

KeyError: 'churn'

## Business Insights
Most customers are retained, but approximately 26.5% have churned, indicating a significant opportunity to improve customer retention.

## Final Business Insights
1. Customer churn is 26.5%, indicating a significant customer-retention challenge.

2. Customers using elctronic check show comparatively high churn, making payment-method behavior an important retention factor.

3. Churned customers generally have lower tenure and lower total charges, suggesting that early-stage customers are more vulnerable to leaving.

4. Monthly charges are higher among churned customers, indicating that pricing or cost sensitivity may contribute to churn.

5. Senior citizens show a relatively higher churn tendency than non-senior customers, suggesting the value of targetred support and retention  offers.

6. The analysis indicates that churn is influenced by multiple customer characteristics rather than a single factor, so retention strategies should be
targeted by customer segment.

7. The business should focus particularly on new customers, high-charge customers, electronic-check users and higher-risk customer segments to improve
retention.



  


 

## Export Final Dataset
Export the final cleaned dataset to a CSV file for use in further analysis and visualization.

In [ ]:
df.to_csv("customer_churn_analysis_cleaned.csv", index=False)